### PR TITLE
[JUJU-2286] Add initial secret backend CLI - add and list

### PR DESCRIPTION
**Pull request by wallyworld**
Commit 1:
Add initial secret backend CLI - add and list
Includes tweaks to facade params to include secret count in the lsit results Secret providers now properly use model config for secret-backend - they use the named provider (this needed some refactoring of the providers)
Model import also needs to account for the mandatory secret-backend model config which defaults to auto auto means use internal (controller) backend for vm models and k8s backend for caas models

Commit 2:
Driveby rename files in secret/provider to split backend and provider code.

Commit 3:
Driveby rename of old secret store method.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

start a vault service, create a config.yaml file with endpoint, token, and unseal key

bootstrap a lxd controller
juju add-secret-backend myvault vault --file /path/to/config.yaml
juju add-model vaultsecrets --config secret-backend=myvault
juju deploy ubuntu
juju exec --unit ubuntu/0 "secret-add foo=bar"

check that the secret is stored in vault

add another model and do not set the secret backend
juju model-config secret-backend -> auto
add a secret and see that it is stored in the controller

juju secret-backends
```
Name      Type        Secrets
internal  controller  666  
myvault   vault       666 
```

bootstrap a k8s model
add a secret and see that the secret is stored in k8s

### CODE REVIEW

File: cmd/juju/secretbackends/add.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	Backend string
+
+	// Attributes from a file.
+	ConfigFile cmd.FileVar
+	// Attributes from key value args.
+	KeyValueAttrs map[string]interface{}
+}
+
+var addSecretBackendsDoc = `
+Displays the secret backends available for storing secret content.
```
Comment (ycliuhw) [inline cmd/juju/secretbackends/add.go:41]: This doc string doesn't seem right.

File: cmd/juju/secretbackends/add.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+)
+
+type addSecretBackendCommand struct {
+	modelcmd.ControllerCommandBase
+	out cmd.Output
+
+	AddSecretBackendsAPIFunc func() (AddSecretBackendsAPI, error)
+
+	Name    string
+	Backend string
```
Comment (ycliuhw) [inline cmd/juju/secretbackends/add.go:32]: For consistency, `BackendType` would be better?
https://github.com/juju/juju/pull/14949/files#diff-f4d1c4834e9a0c8ac391d8157bc4d5f444c04db6f9673853cdbba0f0a8f192f2R13

**Review by ycliuhw — APPROVED** [review]
LGTM, ty

File: apiserver/common/secrets/secrets.go [hunk 3]
```diff
@@ -56,21 +63,35 @@ func ProviderInfoForModel(model Model) (provider.SecretBackendProvider, provider
...
+	case provider.Internal:
+		backendType = juju.BackendType
 	}
-	return GetProvider(backendType)
+
+	if backendType != "" {
+		return GetProvider(backendType)
+	}
+	backendState := GetSecretBackendsState(model)
+	b, err := backendState.GetSecretBackend(backend)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/common/secrets/secrets.go:90]: yes, the default value is always "auto"; it is a required model config value and model config complains if it is missing

File: apiserver/facades/client/secretbackends/secrets.go [hunk 4]
```diff
@@ -56,20 +55,20 @@ func (s *SecretBackendsAPI) createBackend(arg params.SecretBackend) error {
 	if arg.Name == "" {
 		return errors.NotValidf("missing backend name")
 	}
-	if arg.Name == juju.Backend || arg.Name == kubernetes.Backend {
+	if arg.Name == juju.BackendName || arg.Name == "auto" {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/client/secretbackends/secrets.go:58]: yup

File: cmd/juju/secretbackends/add.go (hunk 1 above)
Comment (wallyworld) [reply cmd/juju/secretbackends/add.go:41]: cut and paste error :-)

File: cmd/juju/secretbackends/add.go (hunk 2 above)
Comment (wallyworld) [reply cmd/juju/secretbackends/add.go:32]: yes, ty

**Pull request merged**